x-pack/plugins/security_solution/server/lib/detection_engine/rule_types/eql/create_eql_alert_type.ts

```diff
@@ -32,6 +32,14 @@ export const createEqlAlertType = (
           }
           return validated;
         },
+        validateMutatedParams: (mutatedOject: unknown) => {
+          const mutatedRuleParams = mutatedOject as EqlRuleParams;
+
+          validateImmutable(mutatedRuleParams.immutable);
+          validateIndexPatterns(mutatedRuleParams.index);
+
+          return incrementVersion(mutatedRuleParams);
```

Comment:
it's similar a bit to `validate`, as in validate we pass rule through validator and return new object:
```typescript
          const [validated, errors] = validateNonExact(object, eqlRuleParams);

```
here we can do some asserts and even modify a parameters, knowing that it's called only when rule is mutated: so we have to increment version.
Incrementing version here is better option than in ruleParamsModifier, because you can call `bulkEdit` w/o it, thus skipping version increment

Reply:
I see a couple of drawbacks of that approach. 

First, it creates two places in code where developers could modify rule parameters: `paramsModifier` and `validateMutatedParams`. While `paramsModifier` makes total sense, the second place (`validateMutatedParams`) is not apparent, and developers should keep in mind that the rule mutation code is split into two parts. So to see the whole picture, they should check both. How should developers know about that if there are no clues in the code?

Second, giving the possibility to modify rule parameters after validation kind of kills the idea of validation. So how do we know if the rule after modification inside the validation method is valid?

For example, if I do something like this:

```ts
const validateMutatedParams = (mutatedOject: unknown) => {
  const mutatedRuleParams = mutatedOject as QueryRuleParams;

  validateIndexPatterns(mutatedRuleParams.index); // We've just validated the index value, it's ok

  return incrementVersion({
    ...mutatedRuleParams,
    index: [], // We've mutated the index after validation
  });
},
```

Developers could inadvertently mutate rule objects after the validation. So IMO, if there's a need to change rule parameters after rule modification, it should be done strictly before any validation, and validation functions should not mutate anything passed to them.

Reply:
Yeah, I can see the concerns here.

The idea was to:
Solve for once need of versions increment in every update method by integrating this logic into `validateMutatedParams`

But I'm fine with your approach as well

Reply:
> Solve for once need of versions increment in every update method by integrating this logic into validateMutatedParams

If we need to update the rule version on every change, why don't we implement that invariant on the framework level instead?

Reply:
in terms of alerting perspective: params can be anything, it doesn't care about it.

params is something that enforced on detections level

Reply:
for example: https://github.com/elastic/kibana/blob/main/x-pack/plugins/alerting/server/saved_objects/mappings.ts#L61-L64

Reply:
Don't other solutions need a `version` in their rules? It seems generic enough to be moved to the framework level.

Reply:
Not sure about other solutions, but as I mentioned above, everything within params treated in alerting framework as object and is not validated

Reply:
> everything within params treated in alerting framework as object and is not validated

I'm talking about moving the `version` to the framework level and bumping it automatically on every rule modification. It is not related to validation.

Reply:
sure, but it's not related to this PR

Reply:
I agree this could be addressed separately. As discussed in the team sync a couple  of hours ago, we'll focus only on Security Solution for now.